compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -306,13 +308,52 @@ fn impl_intersection_has_impossible_obligation<'cx, 'tcx>(
     let opt_failing_obligation = [&impl1_header.predicates, &impl2_header.predicates]
         .into_iter()
         .flatten()
-        .map(|&predicate| {
-            Obligation::new(infcx.tcx, ObligationCause::dummy(), param_env, predicate)
+        .map(|&(predicate, span)| {
+            Obligation::new(infcx.tcx, ObligationCause::dummy_with_span(span), param_env, predicate)
         })
         .chain(obligations)
         .find(obligation_guaranteed_to_fail);
 
     if let Some(failing_obligation) = opt_failing_obligation {
+        // Check the failing obligation once again, treating inductive cycles as
+        // ambiguous instead of error.
+        if !infcx.next_trait_solver()
+            && SelectionContext::with_treat_inductive_cycle_as_ambiguous(infcx)
+                .evaluate_root_obligation(&failing_obligation)
+                .map_or(true, |result| result.may_apply())
```

Comment:
:thinking: should we switch the defaults here for 2 reasons?

1. if there are multiple failing obligations, one of which fails because of an inductive cycle, the lint is incorrect
2. perffff: nearly all impls should pass the coherence check, even with ambiguity on inductive cycles. So by first checking with ambiguity on inductive cycles, and then if that fails, checking with error on inductive cycles, should do less work. This also matches how we handle the leak check

Reply:
Yeah ok, makes sense

Reply:
1. should happen only incredibly rarely
2. it's hard to do this, since the new selection context (the one with its mode set to treat-as-ambiguous) would collect the `intercrate_ambiguity_causes`, not the old selection context. we either need to mutate the old selection context to swap its `TreatCycleAs` mode, or expose some API to feed those causes back into the other selection context.